lib/services/access_request.go

```diff
@@ -358,41 +704,151 @@ func (m *RequestValidator) push(role Role) error {
+	astart := len(m.Roles.AllowRequest)
+
+	m.Roles.AllowRequest, err = appendRoleMatchers(m.Roles.AllowRequest, allow.Roles, allow.ClaimsToRoles, m.user.GetTraits())
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	if m.opts.annotate {
+	if m.opts.expandVars {
+		// if this role added additional allow matchers, then we need to record the relationship
+		// between its matchers and its thresholds.  this information is used later to calculate
+		// the rtm and threshold list.
+		if len(m.Roles.AllowRequest) > astart {
+			m.ThresholdMatchers = append(m.ThresholdMatchers, struct {
+				Matchers   []parse.Matcher
+				Thresholds []types.AccessReviewThreshold
+			}{
+				Matchers:   m.Roles.AllowRequest[astart:],
+				Thresholds: allow.Thresholds,
+			})
+		}
+
 		// validation process for incoming access requests requires
 		// generating system annotations to be attached to the request
 		// before it is inserted into the backend.
 		insertAnnotations(m.Annotations.Deny, deny, m.user.GetTraits())
 		insertAnnotations(m.Annotations.Allow, allow, m.user.GetTraits())
+
+		m.SuggestedReviewers = append(m.SuggestedReviewers, allow.SuggestedReviewers...)
 	}
 	return nil
 }
 
+// thresholdCollector is a helper which assembles the Thresholds array for a request.
+// the push() method is used to insert groups of related thresholds and calculate their
+// corresponding index set.
+type thresholdCollector struct {
+	Thresholds []types.AccessReviewThreshold
+}
+
+// push pushes a set of related thresholds and returns the associated indexes.  each set of indexes represents
+// an "or" operator, indicating that one of the referenced thresholds must reach its approval condition in order
+// for the set as a whole to be considered approved.
+func (c *thresholdCollector) push(s []types.AccessReviewThreshold) ([]uint32, error) {
+	if len(s) == 0 {
+		// empty threshold sets are equivalent to the default threshold
+		s = []types.AccessReviewThreshold{
+			{
+				Name:    "default",
+				Approve: 1,
+				Deny:    1,
+			},
+		}
+	}
+
+	var indexes []uint32
+
+	for _, t := range s {
+		tid, err := c.pushThreshold(t)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		indexes = append(indexes, tid)
+	}
+
+	return indexes, nil
+}
+
+// pushThreshold pushes a threshold to the main threshold list and returns its index
+// as a uint32 for compatibility with grpc types.
+func (c *thresholdCollector) pushThreshold(t types.AccessReviewThreshold) (uint32, error) {
+	// maxThresholds is an arbitrary large number that serves as a guard against
+	// odd errors due to casting between int and uint32.  This is probably unnecessary
+	// since we'd likely hit other limitations *well* before wrapping became a concern,
+	// but its best to have explicit guard rails.
+	const maxThresholds = 4096
+
+	// don't bother double-storing equivalent thresholds
+	for i, threshold := range c.Thresholds {
+		if t.Equals(threshold) {
+			return uint32(i), nil
+		}
+	}
```

Comment:
There isn't actually any harm in double-storing (I just figure its good practice not to).  Since threshold names are displayed in the web UI, I don't want teleport to make the decision about which of two descriptions for a given requirement is the one that should be displayed.